l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+% \end{macro}
+% \begin{macro}[pTF]
+%   {
+%     \bitset_if_exist:N, \bitset_if_exist:c
+%   }
+%
+% Existence tests.
+%    \begin{macrocode}
+\prg_new_eq_conditional:NNn
+  \bitset_if_exist:N \str_if_exist:N { p , T , F , TF }
+\prg_new_eq_conditional:NNn
+  \bitset_if_exist:c \str_if_exist:c
+  { p , T , F , TF }
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_set:Nn, \bitset_set:cn,
+%     \bitset_gset:Nn, \bitset_gset:cn,
+%   }
+% A bit is set by either extending the string or by splitting it and
+% then inserting an 1. It is not checked if the value was already 1.
+%    \begin{macrocode}
+% #1 name, #2 index (integer expression,zero based)
+\cs_new_protected:Npn \bitset_set:Nn #1 #2
+  {
+    \int_compare:nNnT {#2 } > { -1 }
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2 + 1 }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 }
+              {
+                \str_put_left:Nn #1 { 0 }
+              }
+            \str_put_left:Nn #1 { 1 }
+          }
+          {
+            %replace value
+            \str_set:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-2 - (#2)}
+                1
+                \str_range:Nnn #1 {0 -(#2)}{-1}
+             }
+          }
+      }
+  }
+
+%#1 name, #2 index (integer expression,zero based)
+\cs_new_protected:Npn \bitset_gset:Nn #1 #2
+  {
+    \int_compare:nNnT {#2 } > { -1 }
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2 + 1 }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 }
+              {
+                \str_gput_left:Nn #1 { 0 }
+              }
+            \str_gput_left:Nn #1 { 1 }
+          }
+          {
+            % replace the value
+            \str_gset:Nx #1
+              {
+                \str_range:Nnn #1 {1}{-2 -(#2)}
+                1
+                \str_range:Nnn #1 {0 -(#2)}{-1}
+              }
+          }
+      }
+  }
+\cs_generate_variant:Nn \bitset_set:Nn  {cn}
+\cs_generate_variant:Nn \bitset_gset:Nn {cn}
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_clear:Nn,  \bitset_clear:cn,
+%     \bitset_gclear:Nn, \bitset_gclear:cn
+%   }
+% Clearing a bit has only to do something if the string is longer than then index.
+%    \begin{macrocode}
+\cs_new_protected:Npn \bitset_clear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
+         {
+           % need to replace the str
+           \str_set:Nx #1
+             {
+               \str_range:Nnn #1 {1}{-2 - (#2)}
+                0
+               \str_range:Nnn #1 { 0 - (#2)}{-1}
+             }
+         }
+     }
+ }
+
+\cs_new_protected:Npn \bitset_gclear:Nn #1 #2  %#1 name, #2 index (zero based)
+ {
+   \int_compare:nNnT {#2 } > { -1 }
+     {
+       \int_compare:nNnT {\str_count:N #1 } > { #2 }
```

Comment:
If `\bitset_clear:Nn` and `\bitset_set:Nn` are roughly the same function, they could be generalised to an internal `\__bitset_set:NNn` that takes `1` or `0` as its first argument.  However `\bitset_clear:Nn` doesn't seem to extend the `bitset var` in case the index `#2` is out of the range.  Is this intentional?

Reply:
Hm, one could even consider a public \bitset_set:NNn, then one would have to find only one name ;-). It is intentional that zeros don't extent the var--imho that isn't needed--but could be easily changed.

Reply:
again based on general bitset usage or on the need for pdf?

Reply:
> It is intentional that zeros don't extent the var

I think either both extend, or both don't.  Mixing just makes it confusing, I think

Reply:
@FrankMittelbach  as far as I got from the bitset package, most uses of bitsets handles them as "numbers" (binary or arabic), and with numbers leading zeros aren't significant.